### PR TITLE
PS-3767: Fixed clang 4/5 warnings for PS 5.7.

### PR DESCRIPTION
**Pull request by dutow**
* Initialized uninitailized variables
* Marked debug variables as possibly unused
* Marked the functions generated by the define_xdr_funcs macro possibly unused. This macro generates several functions, and only some of them is used.
* Removed the unused MYSQLX_PORT global constant
* Corrected two unsigned/signed std::max
* Disabled the unused variable warning for rocksdb, as it's in 3rd party code
* Removed unneccessary assertions from rocksdb (asserting that a non null parameter is non null)
* Added a missing override specifier
* Added additional braces to a struct initalizer
* Changed how optimization is disabled for a function

### CODE REVIEW

File: client/mysql.cc [hunk 1]
```diff
@@ -1430,7 +1430,7 @@ int main(int argc,char *argv[])
     "statement.\n");
   put_info(buff,INFO_INFO);
 
-  uint protocol, ssl_mode;
+  uint protocol = 0, ssl_mode = 0;
```
Comment (laurynas-biveinis) [inline client/mysql.cc]: MySQL coding style: no space before the assignment operator

File: client/mysqladmin.cc [hunk 2]
```diff
@@ -1093,6 +1093,7 @@ static int execute_commands(MYSQL *mysql,int argc, char **argv)
       /* Warn about password being set in non ssl connection */
 #if defined(HAVE_OPENSSL) && !defined(EMBEDDED_LIBRARY)
       uint ssl_mode;
+      ssl_mode = 0;
```
Comment (laurynas-biveinis) [inline client/mysqladmin.cc]: Likewise, and initialize at declaration

File: sql/item_geofunc.cc [hunk 3]
```diff
@@ -1867,7 +1867,7 @@ append_geometry(Geometry::wkb_parser *parser, Json_object *geometry,
         }
         else
         {
-          bool result;
+          bool result = false;
```
Comment (laurynas-biveinis) [inline sql/item_geofunc.cc]: Likewise

File: storage/innobase/include/ut0new.h [hunk 4]
```diff
@@ -663,8 +663,8 @@ class ut_allocator {
 
 		/* e.g. "btr0cur", derived from "/path/to/btr0cur.cc" */
 		char		keyname[FILENAME_MAX];
-		const size_t	len = ut_basename_noext(file, keyname,
-							sizeof(keyname));
+		const size_t	len MY_ATTRIBUTE((unused)) =
```
Comment (laurynas-biveinis) [inline storage/innobase/include/ut0new.h:666]: Why? It's used in ut_a below, which should not be compiled away in any config

File: storage/rocksdb/ha_rocksdb.cc [hunk 5]
```diff
@@ -4210,7 +4210,6 @@ longlong ha_rocksdb::update_hidden_pk_val() {
 
 /* Get the id of the hidden pk id from m_last_rowkey */
 int ha_rocksdb::read_hidden_pk_id_from_rowkey(longlong *const hidden_pk_id) {
```
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc]: IMHO we should remove attribute nonnull instead. Its abuses with nullptrs passed will not be caught by compiler nor generated code, leading to nasty "miscompilations", we had two instances of that in the core server. @georgelorchpercona  ?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
MyRocks changes can be reviewed by either me or @georgelorchpercona, but in this case he has to agree with the plan re. nonnulls first.

File: client/mysqladmin.cc (hunk 2 above)
Comment (dutow) [reply client/mysqladmin.cc]: I wanted to do that, but I couldn't - this part of the function is inside a goto jump (from a later part to an earlier label), and initializing within it is a compilation error.

File: storage/innobase/include/ut0new.h (hunk 4 above)
Comment (dutow) [reply storage/innobase/include/ut0new.h:666]: ut_a is defined to be assert(expr) here, which isn't compiled in release builds.

Comment (laurynas-biveinis) [reply storage/innobase/include/ut0new.h:666]: I see, this must be innochecksum then

**Comment by dutow** [comment]
Updated the PR:
* Added the toku changes (this PR also requires the PerconaFT submodule changes, as the 5.6 branch)
* Removed the non null attributes instaed of the asserts in myrocks

The clang reldebinfo build wasn't fixed, as that build also fails with gcc, and will be addressed in a separate PR.

File: rapid/plugin/group_replication/libmysqlgcs/src/bindings/xcom/gcs_xcom_state_exchange.cc [hunk 6]
```diff
@@ -300,7 +300,7 @@ void Gcs_xcom_state_exchange::reset_with_flush()
 
 void Gcs_xcom_state_exchange::reset()
 {
-  Gcs_xcom_communication_interface *binding_broadcaster=
+  Gcs_xcom_communication_interface *binding_broadcaster MY_ATTRIBUTE((unused)) =
```
Comment (laurynas-biveinis) [inline rapid/plugin/group_replication/libmysqlgcs/src/bindings/xcom/gcs_xcom_state_exchange.cc]: Why not #ifndef NDEBUG instead?

File: sql/item_geofunc.cc (hunk 3 above)
Comment (laurynas-biveinis) [reply sql/item_geofunc.cc]: Still not fixed

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by dutow** [comment]
A small update:

* fixed the previous comments
* rebased to latest 5.7 head
* added the llvm-dev package to the travis config
* fixed two unused function in yassl (for the bundled build)

With these changes, the travis check would be green... except that there are still some errors with gcc, and that it timeouts. Sometimes you can see 1 or 2 green build in it.

File: mysys_ssl/yassl.cc [hunk 7]
```diff
@@ -107,12 +107,12 @@ static int EVP_CipherInit_ex(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
   return 1;
 }
 
-static int EVP_CIPHER_CTX_key_length(const EVP_CIPHER_CTX *ctx)
+MY_ATTRIBUTE((unused)) static int EVP_CIPHER_CTX_key_length(const EVP_CIPHER_CTX *ctx)
```
Comment (laurynas-biveinis) [inline mysys_ssl/yassl.cc]: Please upstream (there consider dropping if they are indeed unused?)

File: testclients/bug25714.c [hunk 8]
```diff
@@ -13,64 +13,52 @@
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
 
+#include <assert.h>
```
Comment (laurynas-biveinis) [inline testclients/bug25714.c]: Non-minimum diff - looks like this file was clang-format'ed, some changes go against MySQL coding standard

File: storage/rocksdb/ha_rocksdb.cc [hunk 9]
```diff
@@ -11463,12 +11463,10 @@ const char *get_rdb_io_error_string(const RDB_IO_ERROR_TYPE err_type) {
   }
 }
 
-// In case of core dump generation we want this function NOT to be optimized
```
Comment (laurynas-biveinis) [inline storage/rocksdb/ha_rocksdb.cc]: Why removed the comment?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysys_ssl/yassl.cc (hunk 7 above)
Comment (dutow) [reply mysys_ssl/yassl.cc]: I changed it to ifndef ndebug

**Comment by dutow** [comment]
Updated based on comments.

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**